### PR TITLE
This is a new chapter that links to the workshop presentation and

### PR DESCRIPTION
**Pull request by gmadrone**
includes much of the content from the workshop.
Signed-off-by: gmadrone <gmadrone@google.com>

### CODE REVIEW

File: implementation/README.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
+The aim of LinuxBoot is to reduce complexity and obscure firmware by moving that functionality into kernel and userspace. This chapter describes the procedures from a [LinuxBoot workshop](https://docs.google.com/presentation/d/1s9ka4v7leKeJa3116AQoNb9cv3OqmnW6pgn0ov9WiHo/edit?ts=5e2b227b#slide=id.g7ceec54197_4_163) where an Atomic Pi board with UEFI firmware was converted to run LinuxBoot. 
+
+## A quick refresher on UEFI
+
+UEFI has three sections:
+
++   SEC ("Boot") 
++   PEI ("Very early chip setup and DRAM programming")
++   DXE ("DRAM code")
+
```
Comment (rminnich) [inline implementation/README.md]: off the top of my head, I'm inclined to leave lines 11-18 out. But let's talk. It's all true and all factual but there's the question of whether we need to hammer UEFI any harder. Line 19 can definitely stay in. Thoughts?

File: implementation/README.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
++   Insert new DXEs
++   Replace the binary code of a DXE with a kernel
+
+If you are lucky, you can always flash with _internal_ flash hardware.
+
+The Management Engine (ME) is an x86 CPU embedded in the Intel Platform Controller Hub (PCH). It runs the Minix operating system which boots first and enables hardware such as clocks and GPIOs. ME checks the contents of flash memory and is used to implement "BootGuard". If you reflash and the ME is in "BootGuard" mode, your machine will be unusable. You need to run a tool called `me_cleaner` on the image to disable BootGuard.
+
+     /usr/bin/python2 me_cleaner.py -s somefile.bin
+
+`me_cleaner` sets the high assurance platform (HAP) bit. HAP is an IT security framework that provides a way to disable a feature on Intel chips that presents a security risk. Setting the bit with `me_cleaner` disables the "feature".  Note that this does not always work; check with your hardware vendor.
```
Comment (rminnich) [inline implementation/README.md]: "presents a security risk" -> "does not allow us to modify the UEFI image and install LinuxBoot"

Again, this is true, based on the many CVEs against the ME,  but ... maybe we don't need to say it.

As for "check with your hardware vendor", hard experience says an answer from the vendors is as useful as a coin flip. 

"check with the linuxboot community."

File: implementation/README.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
++   UEFI Shell (mentioned only for completeness)
+
+    +   Install Linux on FAT-32 media with a name of your choice (e.g. "kernel")
+        +   FAT-32, also known as MS-DOS file system
+    +   Boot kernel at UEFI Shell prompt
+    +   We've run into a few systems that don't have a UEFI shell
+
+#### Working with a system that only has a net interface
+
+If the system only has a net interface, you use Dynamic Host Configuration Protocol (DHCP), using broadcast DISCOVER, and Trivial File Transfer Protocol (TFTP) to get the boot information you need. These programs are provided by the PXEboot program.
```
Comment (rminnich) [inline implementation/README.md]: These programs ... 
is borderline not quite right, I think we can remove this sentence.

File: implementation/README.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
++   After running `utk`, you can reflash
+
+### Step 3: remove as many DXEs as possible
+
++   You can do an initial mass removal based on your current knowledge
++   `utk` automates removing DXEs: this is the DXE cleaner
+
+    +   `utk` removes a DXE, reflashes, checks if it boots, repeats
+
+    This part should be easy: DXE can have a dependency section. In practice, it's hard: because dependency sections are full of errors and omissions. Sloppy UEFI code frequently does not check for failed DXE loads, e.g., remove EnglishDXE, get a General Protection Fault.
```
Comment (rminnich) [inline implementation/README.md]: Sloppy -> A lot of

Comment (rminnich) [reply implementation/README.md]: can remove frequently

Comment (rminnich) [reply implementation/README.md]: DXE loads. -> full stop. you can remove the e.g.

File: implementation/README.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
++   If you can build a DXE from source, you can use `utk` to remove the proprietary one and replace it with one built from source. You can get DXE source from the tianocore/EDK2 source repo at github.com. The GitHub repo has a **_limited_** number of DXEs in source form; i.e., you can't build a full working image using it.
++   There are scripts that let you compile individual DXEs, including the UEFI Shell and Boot Device Selection (BDS). These two DXEs have been compiled and are used in the Atomic Pi. Source-based BDS was needed to ensure the UEFI Shell was called.
++   You only need the UEFI Shell built long enough to replace it with Linux.
+
+### Final step: reflash the image
+
++   "Native" reflash: Boot the system whatever way is easiest: netboot, usb, local disk, and run:
+
+    `flashrom -p internal -w new.bin`
```
Comment (rminnich) [inline implementation/README.md]: where new.bin is a standin a file name which might be anything, what's the right way to say that.

File: implementation/README.md (hunk 1 above)
Comment (gmadrone) [reply implementation/README.md]: I agree. We want to emphasize what users get with LinuxBoot and lines 11-18 don't add to that.

Comment (gmadrone) [reply implementation/README.md]: I deleted lines 11-18.

File: implementation/README.md (hunk 2 above)
Comment (gmadrone) [reply implementation/README.md]: I made these changes.

File: implementation/README.md (hunk 3 above)
Comment (gmadrone) [reply implementation/README.md]: I removed the sentence.

File: implementation/README.md (hunk 4 above)
Comment (gmadrone) [reply implementation/README.md]: I made these three changes.

File: implementation/README.md (hunk 5 above)
Comment (gmadrone) [reply implementation/README.md]: I am addressing the issue in the updated PR. Let me know if it does not work as you'd like.

File: implementation/README.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+1.  Boot from USB stick via UEFI shell command _or_ netboot (zero changes)
+1.  Replace UEFI Shell code section with Linux kernel (change part of one thing)
+1.  Remove as many DXEs as possible (change by removal). This change:
+
+       *  Speeds boot
+       *  Reduces panic possibilities
+       *  Removes exploits
+       *  In production, it has solved problems
+
+1.  Replace some DXEs with open source components (change by replacement)
```
Comment (rminnich) [inline implementation/README.md]: SHould these all be 1?

File: implementation/README.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+There are two tools you use when you modify the UEFI flash image: `utk `and `me_cleaner`
+
+The `utk` tool can:
+
++   Remove DXEs
++   Insert new DXEs
++   Replace the binary code of a DXE with a kernel
+
+If you are lucky, you can always flash with _internal_ flash hardware.
```
Comment (rminnich) [inline implementation/README.md]: This sentence probably does not belong here. It's about flashing.

File: implementation/README.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+The Management Engine (ME) is an x86 CPU embedded in the Intel Platform Controller Hub (PCH). It runs the Minix operating system which boots first and enables hardware such as clocks and GPIOs. ME checks the contents of flash memory and is used to implement "BootGuard". If you reflash and the ME is in "BootGuard" mode, your machine will be unusable. You need to run a tool called `me_cleaner` on the image to disable BootGuard.
+
+ `/usr/bin/python2 me_cleaner.py -s` _imagefile.bin_
+
+`me_cleaner` sets the high assurance platform (HAP) bit. HAP is an IT security framework that provides a way to disable a feature on Intel chips that presents does not allow us to modify the UEFI image and install LinuxBoot. Setting the bit with `me_cleaner` disables the "feature".  Note that this does not always work; check with the LinuxBoot community.
```
Comment (rminnich) [inline implementation/README.md]: HAP is an IT security framework that provides -> HAP provides

Comment (rminnich) [reply implementation/README.md]: "presents" -> remove that word

File: implementation/README.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+If the system only has a net interface, you use Dynamic Host Configuration Protocol (DHCP), using broadcast DISCOVER, and Trivial File Transfer Protocol (TFTP) to get the boot information you need. 
+
+Configuration information is provided by REPLY to a DHCP request. The REPLY returns an IP, server, and a configuration file name that provides:
+
+   +   Identity
+   +   What to boot
+   +   Where to get it
+
+Data is provided by TFTP. On proper systems, HTTP downloading takes a fraction of a second even for 16M kernels. With TFTP it's very slow and TFTP won't work with initramfs much large than 32MiB. Most LinuxBoot shops use or are transitioning to HTTP.
```
Comment (rminnich) [inline implementation/README.md]: "On proper systems, " -> can remove that text

File: implementation/README.md (hunk 6 above)
Comment (gmadrone) [reply implementation/README.md]: Yes, in markdown, we use 1 for all steps and it generates the correct numbering when it is rendered.

File: implementation/README.md (hunk 7 above)
Comment (gmadrone) [reply implementation/README.md]: Removed.

File: implementation/README.md (hunk 8 above)
Comment (gmadrone) [reply implementation/README.md]: Fixed these two.

File: implementation/README.md (hunk 9 above)
Comment (gmadrone) [reply implementation/README.md]: Removed.

**Pull request merged**